tethys_apps/base/controller.py

```diff
@@ -18,3 +48,286 @@ def as_controller(cls, **kwargs):
         Thin veneer around the as_view method to make interface more consistent with Tethys terminology.
         """
         return cls.as_view(**kwargs)
+
+
+def controller(
+        function_or_class: Union[Callable[[HttpRequest, ...], Any], TethysController] = None,
+        /, *,
+
+        # UrlMap Overrides
+        name: str = None,
+        url: Union[str, list, tuple, dict, None] = None,
+        protocol: str = 'http',
+        regex: Union[str, list, tuple] = None,
+        handler: Union[str, Callable] = None,
+        handler_type: str = None,
+
+        # login_required kwargs
+        login_required: bool = True,
+        redirect_field_name: str = REDIRECT_FIELD_NAME,
+        login_url: str = None,
+
+        # workspace decorators
+        app_workspace: bool = False,
+        user_workspace: bool = False,
+
+        # ensure_oauth2 kwarg
+        ensure_oauth2_provider: str = None,
+
+        # enforce_quota kwargs
+        enforce_quotas: Union[str, list, tuple, None] = None,
+
+        # permission_required kwargs
+        permissions_required: Union[str, list, tuple] = None,
+        permissions_use_or: bool = False,
+        permissions_message: str = None,
+        permissions_raise_exception: bool = False,
+
+        # additional kwargs to pass to TethysController.as_controller
+        **kwargs
+) -> Callable:
+    """
+    Decorator to register a function or TethysController class as a controller
+    (by automatically registering a UrlMap for it).
+
+    Args:
+        name: Name of the url map. Letters and underscores only (_). Must be unique within the app. The default is the name of the function being decorated.
```

Comment:
What if you have two functions with the same name in different submodules of controllers?

* controllers.resource.view
* controllers.workflows.view

Reply:
You would have to manually provide a name for at least one of them.

Reply:
I'm now catching that case and providing a different name, but printing out a big warning message:

![image](https://user-images.githubusercontent.com/6454472/156667446-27ec2834-7d7b-40cf-8dbb-567f9e87266d.png)